ethereumetl/jobs/export_geth_traces_job.py

```diff
@@ -0,0 +1,79 @@
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import json
+
+from ethereumetl.executors.batch_work_executor import BatchWorkExecutor
+from ethereumetl.json_rpc_requests import generate_trace_block_by_number_json_rpc
+from ethereumetl.jobs.base_job import BaseJob
+from ethereumetl.mappers.geth_trace_mapper import EthGethTraceMapper
+from ethereumetl.utils import validate_range, rpc_response_batch_to_results
+
+
+# Exports geth traces
+class ExportGethTracesJob(BaseJob):
+    def __init__(
+            self,
+            start_block,
+            end_block,
+            batch_size,
+            batch_web3_provider,
+            max_workers,
+            item_exporter):
+        validate_range(start_block, end_block)
+        self.start_block = start_block
+        self.end_block = end_block
+
+        self.batch_web3_provider = batch_web3_provider
+
+        self.batch_work_executor = BatchWorkExecutor(batch_size, max_workers)
+        self.item_exporter = item_exporter
+
+        self.geth_trace_mapper = EthGethTraceMapper()
+
+    def _start(self):
+        self.item_exporter.open()
+
+    def _export(self):
+        self.batch_work_executor.execute(
+            range(self.start_block, self.end_block + 1),
+            self._export_batch,
+            total_items=self.end_block - self.start_block + 1
+        )
+
+    def _export_batch(self, block_number_batch):
+        blocks_rpc = list(generate_trace_block_by_number_json_rpc(block_number_batch))
+        response = self.batch_web3_provider.make_request(json.dumps(blocks_rpc))
+        results = rpc_response_batch_to_results(response)
+
+        # adding block_numbers here, as RPC response doesn't have them
+        geth_traces = [self.geth_trace_mapper.json_dict_to_geth_trace({
+            'block_number': block_number,
+            'traces': traces,
+        }) for block_number, traces in zip(block_number_batch, results)]
```

Comment:
Not sure we can assume the responses will come in the same order as requests. As a workaround block number can be put to JSON RPC request id, then it will be in the id field in the response.

Reply:
Refactored this to put (and get) block number in request ID.